### PR TITLE
destructor regression fixes

### PR DESCRIPTION
**Pull request by cooldome**
Recent `injectdestructors.nim` refactoring caused a number of regressions for my code base.
Though I had hard time to reproduce issues in the small test cases. 
Regressions appear in most complicated cases.

No tests I am afraid. I provide comments per line instead

### CODE REVIEW

File: compiler/injectdestructors.nim [hunk 1]
```diff
@@ -471,10 +472,9 @@ proc recurse(n: PNode, c: var Con, processProc: proc): PNode =
           branch.add processProc(n[i][0], c)
       result.add branch
   else:
-    assert(false, $n.kind)
```
Comment (cooldome) [inline compiler/injectdestructors.nim]: I got nodes of kind nkRefTy, nkRaiseStmt and nkTypeOfExpr occured here.

File: compiler/injectdestructors.nim [hunk 2]
```diff
@@ -471,10 +472,9 @@ proc recurse(n: PNode, c: var Con, processProc: proc): PNode =
...
+    result = copyTree(n)
 
 proc pExpr(n: PNode; c: var Con): PNode =
-  assert(isExpression(n), $n.kind)
```
Comment (cooldome) [inline compiler/injectdestructors.nim]: nkCall with a function that returns nothing or void `isExpression` returns false while nkCall is properly handled further on

Comment (Clyybber) [reply compiler/injectdestructors.nim]: How does such a function end up here?

Comment (cooldome) [reply compiler/injectdestructors.nim]: Compile with --newruntime and --threads:on to trigger this assert:
```nim
import threadpool

proc test(a: string) = 
  echo a

spawn test("my")

```

File: compiler/injectdestructors.nim (hunk 1 above)
Comment (Clyybber) [reply compiler/injectdestructors.nim]: I'd prefer handling them explicitly and leaving the assert in there

File: compiler/injectdestructors.nim (hunk 2 above)
Comment (Clyybber) [reply compiler/injectdestructors.nim]: Can we add this as a test?

**Comment by Araq** [comment]
Closing in favor of https://github.com/nim-lang/Nim/pull/12169

File: compiler/injectdestructors.nim [hunk 3]
```diff
@@ -412,7 +412,6 @@ template isExpression(n: PNode): bool =
   (not isEmptyType(n.typ)) or (n.kind in nkLiterals + {nkNilLit, nkRange})
 
 proc recurse(n: PNode, c: var Con, processProc: proc): PNode =
-  if n.sons.len == 0: return n
```
Comment (Clyybber) [inline compiler/injectdestructors.nim]: @cooldome I'm redoing the refactoring. Can you explain why you had to have this line removed?

**Pull request closed without merging**